lib/winapi/sync.c

```diff
@@ -38,6 +174,244 @@ VOID LeaveCriticalSection (LPCRITICAL_SECTION lpCriticalSection)
     RtlLeaveCriticalSection(lpCriticalSection);
 }
 
+#define SRW_GLOBAL_LOCK_MASK ((uintptr_t)0x40000000)
+#define SRW_READER_LOCK_MASK ((uintptr_t)0x80000000)
+#define SRW_READER_COUNT_MASK ~(SRW_GLOBAL_LOCK_MASK | SRW_READER_LOCK_MASK)
+
+void AcquireSRWLockExclusive (PSRWLOCK SRWLock)
+{
+    // Grab the global lock (second-highest bit)
+    uintptr_t unlocked_val = SRWLock->Ptr & ~SRW_GLOBAL_LOCK_MASK;
+    while (!__sync_bool_compare_and_swap(&SRWLock->Ptr, unlocked_val, unlocked_val | SRW_GLOBAL_LOCK_MASK)) {
+        SwitchToThread();
+        unlocked_val = SRWLock->Ptr & ~SRW_GLOBAL_LOCK_MASK;
+    }
+}
+
+void AcquireSRWLockShared (PSRWLOCK SRWLock)
+{
+    // Grab the reader lock
+    uintptr_t unlocked_val = SRWLock->Ptr & ~SRW_READER_LOCK_MASK;
+    while (!__sync_bool_compare_and_swap(&SRWLock->Ptr, unlocked_val, unlocked_val | SRW_READER_LOCK_MASK)) {
+        SwitchToThread();
+        unlocked_val = SRWLock->Ptr & ~SRW_READER_LOCK_MASK;
+    }
+
+    // Increase reader count by one
+    uintptr_t prev_val = __atomic_fetch_add(&SRWLock->Ptr, 1, __ATOMIC_ACQ_REL);
+    if ((prev_val & SRW_READER_COUNT_MASK) == 0) {
+        // If we're the first reader, grab the global lock, too
+        AcquireSRWLockExclusive(SRWLock);
+    }
+
+    // Release the reader lock
+    __sync_fetch_and_and(&SRWLock->Ptr, ~SRW_READER_LOCK_MASK);
+}
+
+void InitializeSRWLock (PSRWLOCK SRWLock)
+{
+    __atomic_store_n(&SRWLock->Ptr, 0, __ATOMIC_RELEASE);
+}
+
+void ReleaseSRWLockExclusive (PSRWLOCK SRWLock)
+{
+    // Release the global lock (second-highest bit)
+    __sync_fetch_and_and(&SRWLock->Ptr, ~SRW_GLOBAL_LOCK_MASK);
+}
+
+void ReleaseSRWLockShared (PSRWLOCK SRWLock)
+{
+    // Grab the reader lock
+    uintptr_t unlocked_val = SRWLock->Ptr & ~SRW_READER_LOCK_MASK;
+    while (!__sync_bool_compare_and_swap(&SRWLock->Ptr, unlocked_val, unlocked_val | SRW_READER_LOCK_MASK)) {
+        SwitchToThread();
+        unlocked_val = SRWLock->Ptr & ~SRW_GLOBAL_LOCK_MASK;
+    }
+
+    // Decrease reader count by one
+    uintptr_t prev_val = __atomic_fetch_sub(&SRWLock->Ptr, 1, __ATOMIC_ACQ_REL);
+    if ((prev_val & SRW_READER_COUNT_MASK) == 1) {
```

Comment:
nit: I might be a bit paranoid, but we could consider having `assert((prev_val & SRW_READER_COUNT_MASK) != 0)` to avoid people from releasing non-owned locks.

Reply:
Added.